shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm

```diff
@@ -361,77 +428,182 @@
   composition_order_.clear();
   active_composition_order_.clear();
   picture_recorders_.clear();
+  platform_view_rtrees_.clear();
   current_composition_params_.clear();
   clip_count_.clear();
   views_to_recomposite_.clear();
+  layer_pool_->RecycleLayers();
+}
+
+SkRect FlutterPlatformViewsController::GetPlatformViewRect(int view_id) {
+  UIView* platform_view = [views_[view_id].get() view];
+  UIScreen* screen = [UIScreen mainScreen];
+  CGRect platform_view_cgrect = [platform_view convertRect:platform_view.bounds
+                                                    toView:flutter_view_];
+  return SkRect::MakeXYWH(platform_view_cgrect.origin.x * screen.scale,    //
+                          platform_view_cgrect.origin.y * screen.scale,    //
+                          platform_view_cgrect.size.width * screen.scale,  //
+                          platform_view_cgrect.size.height * screen.scale  //
+  );
 }
 
 bool FlutterPlatformViewsController::SubmitFrame(GrContext* gr_context,
-                                                 std::shared_ptr<IOSContext> ios_context) {
+                                                 std::shared_ptr<IOSContext> ios_context,
+                                                 SkCanvas* background_canvas) {
   DisposeViews();
-
-  bool did_submit = true;
-  for (int64_t view_id : composition_order_) {
-    EnsureOverlayInitialized(view_id, ios_context, gr_context);
-    auto frame = overlays_[view_id]->surface->AcquireFrame(frame_size_);
-    // If frame is null, AcquireFrame already printed out an error message.
-    if (frame) {
-      SkCanvas* canvas = frame->SkiaCanvas();
-      canvas->drawPicture(picture_recorders_[view_id]->finishRecordingAsPicture());
-      canvas->flush();
-      did_submit &= frame->Submit();
+  // Clipping the background canvas before drawing the picture recorders requires to
+  // save and restore the clip context.
+  SkAutoCanvasRestore save(background_canvas, /*doSave=*/true);
+  // Maps a platform view id to a vector of `FlutterPlatformViewLayer`.
+  LayersMap platform_view_layers;
+
+  auto did_submit = true;
+  auto num_platform_views = composition_order_.size();
+
+  for (size_t i = 0; i < num_platform_views; i++) {
+    int64_t platform_view_id = composition_order_[i];
+    sk_sp<RTree> rtree = platform_view_rtrees_[platform_view_id];
+    sk_sp<SkPicture> picture = picture_recorders_[platform_view_id]->finishRecordingAsPicture();
```

Comment:
This is one of the places where there were unchecked indexing happening.  How are you confident that this won't be a problem?  Seems like we should at the minimum have a FML_DCHECK here.

Reply:
Every entry in `composition_order_` always has an entry in `picture_recorders_` and `platform_view_rtrees_`.  This is currently tested in the scenario app.